tests/api/v1/subjects/putWithParent.js

```diff
@@ -0,0 +1,308 @@
+    .set('Authorization', token)
+    .send({
+      name: `${tu.namePrefix}Canada`,
+      parentId: i0,
+      parentAbsolutePath: n0.name,
+      isPublished: true,
+    })
+    .expect(constants.httpStatus.OK)
+    .end((err  , res ) => {
+      if (err) {
+        done(err);
+      }
+
+      expect(res.body.parentAbsolutePath).to.equal(n0.name);
+      expect(res.body.parentId).to.equal(i0);
+      done();
+    });
+  });
+
+  it('fail if parentAbsolutePath refers to a different subject than specified by parentId', (done) => {
+    api.put(`${path}/${i0}`)
+    .set('Authorization', token)
+    .send({
+      name: `${tu.namePrefix}Canada`,
+      parentId: i0a,
+      parentAbsolutePath: _root.name,
+      isPublished: true,
+    })
+    .expect(constants.httpStatus.BAD_REQUEST)
+    .end((err  , res ) => {
+      if (err) {
+        done(err);
+      }
+
+      expect(res.body.errors[0].type).to.equal('ParentSubjectNotMatch');
+      done();
+    });
+  });
+
+  it('fail if parentAbsolutePath refers to non-existent subject', (done) => {
+    api.put(`${path}/${i0}`)
+    .set('Authorization', token)
+    .send({
+      name: `${tu.namePrefix}Canada`,
+      parentId: i0a,
+      parentAbsolutePath: 'iDontExist',
+      isPublished: true,
+    })
+    .expect(constants.httpStatus.BAD_REQUEST)
+    .end((err  , res ) => {
+      if (err) {
+        done(err);
+      }
+
+      expect(res.body.errors[0].type).to.equal('ParentSubjectNotFound');
+      done();
+    });
+  });
+
+  it('sets the parent based on the parentId', (done) => {
+    api.put(`${path}/${i1}`)
+    .set('Authorization', token)
+    .send({
+      name: p1.name,
+      isPublished: p1.isPublished,
+      parentId: i0a,
+    })
+    .expect(constants.httpStatus.OK)
+    .end((err, res ) => {
+      if (err) {
+        done(err);
+      }
+
+      expect(res.body.parentId).to.equal(i0a);
+      expect(res.body.parentAbsolutePath).to.equal(n0a.name);
+      done();
+    });
+  });
+
+  it('sets the parent based on the parentAbsolutePath', (done) => {
+    const NEW_NAME = 'newName';
+    api.put(`${path}/${i0}`)
+    .set('Authorization', token)
+    .send({
+      name: NEW_NAME,
+      isPublished: p1.isPublished,
+      parentAbsolutePath: _root.name,
+    })
+    .expect(constants.httpStatus.OK)
+    .end((err, res ) => {
+      if (err) {
+        done(err);
+      }
+
+      expect(res.body.parentId).to.equal(iRoot);
+      expect(res.body.parentAbsolutePath).to.equal(_root.name);
+      done();
+    });
+  });
+
+  it('with NEITHER parentId NOR parentAbsolutePath, reparent the subject as a root subject', (done) => {
+    const NAME = 'iAmRoot';
+    api.put(`${path}/${i0}`)
+    .set('Authorization', token)
+    .send({ name: NAME, isPublished: true })
+    .expect(constants.httpStatus.OK)
+    .end((err  , res ) => {
+      if (err) {
+        done(err);
+      }
+
+      expect(res.body.absolutePath).to.equal(NAME);
+      done();
+    });
+  });
```

Comment:
Test is failing:

  it.only('on un-publish, does not re-parent', (done) => {
    api.put(`${path}/${i1}`)
    .set('Authorization', token)
    .send({
      name: `${tu.namePrefix}Canada`,
      isPublished: false,
    })
    .expect(constants.httpStatus.OK)
    .end((err  , res ) => {
      if (err) {
        done(err);
      }

      console.log(res.body)
      expect(res.body.parentId).to.equal(i0); // parentId field is undefined
      expect(res.body.parentAbsolutePath).to.equal(n0.name); // '' instead of non-empty value: shows subject is re-parented
      done();
    });
  });

I am looking at the culprit files in db. Will push another commit when that is fixed